metals/src/main/scala/scala/meta/internal/metals/Diagnostics.scala

```diff
@@ -98,16 +98,20 @@ final class Diagnostics(
         syntaxError(path) = diagnostic
         publishDiagnostics(path)
       case _ =>
-        onNoSyntaxError(path)
+        onClose(path)
     }
   }
 
-  def onNoSyntaxError(path: AbsolutePath): Unit = {
-    syntaxError.remove(path) match {
-      case Some(_) =>
+  def onClose(path: AbsolutePath): Unit = {
+    val diags = if (path.isWorksheet) {
+      diagnostics.remove(path).toList.flatMap(_.asScala) ++
+        syntaxError.remove(path)
+    } else syntaxError.remove(path).toList
+    diags match {
+      case Nil =>
+        () // Do nothing, there was no previous error.
+      case _ =>
```

Comment:
`onClose` called from `onSyntaxError` seems a bit strange. What do you think about:
- adding some private method called e.g. `clearDiagostiscFor`
- `onClose` and `case _  =>` in `onSyntaxError` can call `clearDiagostiscFor`
- `clearDiagostiscFor` will have scaladoc which will explain where it's used

Reply:
Och, forgot about the comment. I will change that in another PR